ledger/eval.go

```diff
@@ -92,32 +92,145 @@ func (x *roundCowBase) blockHdr(r basics.Round) (bookkeeping.BlockHeader, error)
 	return x.l.BlockHdr(r)
 }
 
+func (x *roundCowBase) allocated(addr basics.Address, aidx basics.AppIndex, global bool) (bool, error) {
+	acct, _, err := x.l.LookupWithoutRewards(x.rnd, addr)
+	if err != nil {
+		return false, err
+	}
+
+	// For global, check if app params exist
+	if global {
+		_, ok := acct.AppParams[aidx]
+		return ok, nil
+	}
+
+	// Otherwise, check app local states
+	_, ok := acct.AppLocalStates[aidx]
+	return ok, nil
+}
+
+// getKey gets the value for a particular key in some storage
+// associated with an application globally or locally
+func (x *roundCowBase) getKey(addr basics.Address, aidx basics.AppIndex, global bool, key string) (basics.TealValue, bool, error) {
```

Comment:
It is getting the value. Why not call it getValue?

Reply:
It must be called `getKey` because it is part of `roundCowParent` interface. `GetKey`/`SetKey` looked more appropriate than `GetValue`/`SetValue`, and `roundCowState.getKey` is implementation of `GetKey`, and both `roundCowState` and `roundCowBase` implement `roundCowParent` interface.

Reply:
okay :-)